### PR TITLE
refactor: avoid cross-package relative Sass imports to support Yarn PnP

### PR DESCRIPTION
**Pull request by devversion**
Fixes https://github.com/angular/components/issues/24128

See individual commits. Note that this is not necessarily only being about Yarn 2.x+, but can also
be considered a general cleanup, enforcing consistent ways of importing code between packages.

Also, in the future, deep cross-NPM package relative imports might break regardless (if not explicitly specified
in the `package.json#exports` - adding all files is counterintuitive). We are currently relying on unspecified behavior that can change at any
time when e.g. `webpack` sass-loader changes, or when NPM packages are laid out differently etc.

As part of this change, I cleaned-up most missing Sass dependencies. One downside of this change is
that most Sass files now rely on the larger bulk `_index.scss` files (and its transtive libs) (see note #2 in details)

<details>

**Note #1**: For Yarn PnP support/or our NPM packages to be future-proof/valid in respect to package exports, it would have been sufficient to only update the theming Sass files. I've decided to also apply the changes to our component Sass files (sass binaries) since that will make the importing consistent/easier to remember (and the lint rule also)

**Note #2**: This will increase the action input for some Sass binaries, but that seems acceptable given that all unit tests
rely on a prebuilt-theme anyway etc. This will also match with what NPM users are seeing. Certainly a little
unfortunate to give-up on a little fine-grained targets here but we haven't been very consistent with that anyay.. (and it's a trade-off)
</details>

### CODE REVIEW

**Review by mmalerba — APPROVED** [review]
LGTM

File: src/material/_index.scss [hunk 1]
```diff
@@ -14,6 +14,25 @@
 @forward './core/typography/typography-utils' show typography-level,
   font-size, line-height, font-weight, letter-spacing, font-family, font-shorthand;
 
+// Private/Internal
+@forward './core/density/private/all-density' as private-* show private-all-component-densities;
```
Comment (crisbeto) [inline src/material/_index.scss:18]: Instead of exposing these through the main entry point, should we have a separate entry point instead?

Comment (devversion) [reply src/material/_index.scss:18]: Yes, I could do that, but it has two consequences:

* Need for maintaining/wiring up another file in the NPM package (e.g. in the `exports`)
* Increase in inconsistency for cross-package imports (where currently there is just a single way)

Also people might still import from it. I believe there is not much benefit, or do you have other thoughts?

**Comment by andrewseguin** [comment]
Needs rebase - this needs some special care getting into Google. Let me know the next time you have this rebased so I can focus on it

**Comment by andrewseguin** [comment]
Sorry @devversion - needs another little rebase. I think I've got the internal stuff almost working so it shouldn't be much longer

**Pull request merged**